src/bootstrap/bootstrap.py

```diff
@@ -0,0 +1,316 @@
+        start = line.find('"')
+        end = start + 1 + line[start+1:].find('"')
+        return line[start+1:end]
+
+    def exe_suffix(self):
+        if sys.platform == 'win32':
+            return '.exe'
+        else:
+            return ''
+
+    def parse_nightly_dates(self):
+        nightlies = os.path.join(self.rust_root, "src/nightlies.txt")
+        with open(nightlies, 'r') as nightlies:
+            rustc, cargo = nightlies.read().split("\n")[:2]
+            assert rustc.startswith("rustc: ")
+            assert cargo.startswith("cargo: ")
+            self._rustc_date = rustc[len("rustc: "):]
+            self._cargo_date = cargo[len("cargo: "):]
+
+    def build_bootstrap(self):
+        env = os.environ.copy()
+        env["CARGO_TARGET_DIR"] = os.path.join(self.build_dir, "bootstrap")
+        env["RUSTC"] = self.rustc()
+        env["LD_LIBRARY_PATH"] = os.path.join(self.bin_root(), "lib")
+        env["DYLD_LIBRARY_PATH"] = os.path.join(self.bin_root(), "lib")
+        env["PATH"] = os.path.join(self.bin_root(), "bin") + \
+                      os.pathsep + env["PATH"]
+        self.run([self.cargo(), "build", "--manifest-path",
+                  os.path.join(self.rust_root, "src/bootstrap/Cargo.toml")],
+                 env)
+
+    def run(self, args, env):
+        proc = subprocess.Popen(args, env = env)
+        ret = proc.wait()
+        if ret != 0:
+            sys.exit(ret)
+
+    def build_triple(self):
+        config = self.get_toml('build')
+        if config:
+            return config
+        config = self.get_mk('CFG_BUILD')
+        if config:
+            return config
+        try:
+            ostype = subprocess.check_output(['uname', '-s']).strip()
+            cputype = subprocess.check_output(['uname', '-m']).strip()
+        except FileNotFoundError:
+            if sys.platform == 'win32':
+                return 'x86_64-pc-windows-msvc'
+            else:
+                raise
+
+        # Darwin's `uname -s` lies and always returns i386. We have to use
+        # sysctl instead.
+        if ostype == 'Darwin' and cputype == 'i686':
+            sysctl = subprocess.check_output(['sysctl', 'hw.optional.x86_64'])
+            if sysctl.contains(': 1'):
+                cputype = 'x86_64'
+
+        # The goal here is to come up with the same triple as LLVM would,
+        # at least for the subset of platforms we're willing to target.
+        if ostype == 'Linux':
+            ostype = 'unknown-linux-gnu'
+        elif ostype == 'FreeBSD':
+            ostype = 'unknown-freebsd'
+        elif ostype == 'DragonFly':
+            ostype = 'unknown-dragonfly'
+        elif ostype == 'Bitrig':
+            ostype = 'unknown-bitrig'
+        elif ostype == 'OpenBSD':
+            ostype = 'unknown-openbsd'
+        elif ostype == 'NetBSD':
+            ostype = 'unknown-netbsd'
+        elif ostype == 'Darwin':
+            ostype = 'apple-darwin'
+        elif ostype.startswith('MINGW'):
```

Comment:
So if I run this from Windows command prompt, I end up with `pc-windows-msvc`, but if I run it from MSYS shell, I get `pc-windows-gnu`?  Feels a bit too subtle.
Besides, neither one guarantees that the appropriate compiler is actually installed.  I think I'd rather set default to one of these targets (perhaps `-gnu` since that's what the current build system does), and have user explicitly specify the other via the `--build=` flag.